File: src/dmrgpy/pyitensor/tebd.py
"""TEBD: 2nd-order Trotter (odd/even bond) real-time evolution for
strictly nearest-neighbor Hamiltonians.

Unlike tdvp.py's per-bond, per-step Krylov exponentiation of an
environment-projected *effective* Hamiltonian, each bond's evolution gate
exp(-i*tau*h_bond) here is the exact exponential of the *bare* local
2-site Hamiltonian (a small dense matrix, exponentiated once via
scipy.linalg.expm) -- the standard Vidal TEBD algorithm. Since H is
time-independent for every caller in this codebase (a "quench" holds H
fixed across all nt steps), every gate is built once at setup and reused
unchanged for every subsequent time step: no per-step diagonalization at
all, only a gate contraction + SVD truncation per bond per step. This is
what makes TEBD cheaper than TDVP for models where it applies, at the
cost of only supporting strictly nearest-neighbor H (any term spanning 3+
distinct sites raises NotImplementedError -- long-range Hamiltonians need
TDVP or the MPO-Taylor path instead).

One-site (onsite) terms are split half-and-half onto each of a site's
neighboring bonds (full weight at a chain boundary, where a site has only
one neighbor) -- the same convention TeNPy's NearestNeighborModel uses.
Bare scalar terms (no site reference at all, e.g. a ground-state-energy
shift folded in as `coef*Identity`) are absorbed entirely into bond 1;
since they act as Identity on every other bond too, which bond carries
them doesn't matter.

Fermionic terms need no special-casing here beyond what HTerm.resolve()
already does: for any *physical* (fermion-parity-conserving) Hamiltonian
term restricted to two adjacent sites, resolve()'s Jordan-Wigner carry
always returns to its initial (False) parity by the far site, so every
site beyond the term's own span gets a plain Id, not a trailing F string --
i.e. embedding kron(mats[lo-1], mats[hi-1]) with implicit identity
elsewhere reproduces exactly the same operator the true (JW-threaded) MPO
Hamiltonian represents. bond_hamiltonians() asserts this explicitly rather
than trusting it silently, so a non-parity-conserving term (which would
need a JW string extending past the two-site bond, incompatible with a
local gate) fails loudly instead of silently producing wrong physics.
"""

import numpy as np
from scipy.linalg import expm

from .autompo import AutoMPO
from .mpscontainer import _link_at
from .svd import svd
from .tensor import ITensor, noPrime


def _term_span(term):
    sites = [site for _, site in term.ops]
    if not sites:
        return None
    return min(sites), max(sites)


def bond_hamiltonians(sites, terms):
    """terms: dmrgpy MultiOperator.to_terms() shape, i.e.
    [(coef, [(opname,site), ...]), ...]. Returns {b: (D,D) ndarray} for
    b=1..n-1 (the bond between sites b and b+1), in the same standard
    (out,in) convention as HTerm.resolve()."""
    n = sites.length()
    if n < 2:
        raise ValueError("bond_hamiltonians: TEBD needs at least 2 sites")
    ampo = AutoMPO.from_terms(sites, terms)
    dims = [sites.dim(i) for i in range(1, n + 1)]
    H = {b: np.zeros((dims[b - 1] * dims[b], dims[b - 1] * dims[b]), dtype=complex)
         for b in range(1, n)}

    for term in ampo.terms:
        span = _term_span(term)
        if span is None:
            H[1] += term.coef * np.eye(dims[0] * dims[1], dtype=complex)
            continue
        lo, hi = span
        if hi - lo > 1:
            raise NotImplementedError(
                "TEBD requires a nearest-neighbor Hamiltonian; found a term "
                "spanning sites {}..{}".format(lo, hi))
        mats = term.resolve(sites)  # standard (out,in) per-site matrices
        if hi < n:
            beyond = mats[hi]  # site hi+1 (0-based index hi), just past the span
            ident = np.eye(dims[hi], dtype=complex)
            if not np.allclose(beyond, ident):
                raise NotImplementedError(
                    "TEBD requires fermion-parity-conserving terms (a "
                    "non-parity-conserving term needs a Jordan-Wigner "
                    "string reaching past sites {}..{}, incompatible with "
                    "a local bond gate)".format(lo, hi))
        if hi == lo:
            local = term.coef * mats[lo - 1]
            targets = []
            if lo > 1:
                targets.append((lo - 1, "right"))  # bond (lo-1,lo): term is the right site
            if lo < n:
                targets.append((lo, "left"))       # bond (lo,lo+1): term is the left site
            weight = 0.5 if len(targets) == 2 else 1.0
            for b, side in targets:
                if side == "right":
                    other_dim = dims[b - 1]
                    piece = np.kron(np.eye(other_dim, dtype=complex), local)
                else:
                    other_dim = dims[b]
                    piece = np.kron(local, np.eye(other_dim, dtype=complex))
                H[b] += weight * piece
        else:
            H[lo] += term.coef * np.kron(mats[lo - 1], mats[hi - 1])
    return H


def _bond_gate(sites, i, h_mat, tau):
    """The 2-site ITensor gate exp(-i*tau*h_mat) for bond (i,i+1)."""
    s_i, s_j = sites.si(i), sites.si(i + 1)
    d_i, d_j = s_i.dim, s_j.dim
    U = expm(-1j * tau * h_mat)
    stored = U.T.reshape(d_i, d_j, d_i, d_j)  # (in_i,in_j,out_i,out_j)
    return ITensor((s_i, s_j, s_i.prime(1), s_j.prime(1)), stored)


def _apply_bond_gate(psi, i, gate, cutoff, maxdim):
    """Contract `gate` onto bond (i,i+1) and SVD-truncate. Canonicalizes
    psi's orthogonality center to site i first (via position(), the same
    primitive autompo.py/mpsalgebra.py use for their own truncating
    sweeps), which is what makes the local SVD truncation-optimal
    regardless of which bond was last touched. Mutates psi in place."""
    psi.position(i, cutoff=cutoff, maxdim=maxdim)
    left_link = _link_at(psi, i, i - 1)
    s_i = next(ind for ind in psi.A(i).inds if ind.hastags("Site"))
    theta = noPrime(gate * (psi.A(i) * psi.A(i + 1)), "Site")
    left_inds = ([left_link] if left_link else []) + [s_i]
    U, S, V, _spec = svd(theta, left_inds, cutoff=cutoff, maxdim=maxdim)
    psi.set_A(i, U)
    psi.set_A(i + 1, S * V)
    psi.center = i + 1


class TEBDEvolver:
    """Precomputes every bond's dt/2 and dt evolution gates once from a
    (static, nearest-neighbor) Hamiltonian, then applies a 2nd-order
    Trotter step exp(-i dt/2 H_odd) exp(-i dt H_even) exp(-i dt/2 H_odd)
    per call to step() -- H_odd/H_even being the sums over odd-indexed
    and even-indexed bonds respectively, each internally commuting since
    every bond in one group acts on disjoint sites."""

    def __init__(self, sites, terms, dt, cutoff, maxdim):
        self.sites = sites
        self.n = sites.length()
        self.cutoff = cutoff
        self.maxdim = maxdim
        h_bonds = bond_hamiltonians(sites, terms)
        self._gates_half = {b: _bond_gate(sites, b, h, dt / 2.0) for b, h in h_bonds.items()}
        self._gates_full = {b: _bond_gate(sites, b, h, dt) for b, h in h_bonds.items()}
        self._odd = list(range(1, self.n, 2))
        self._even = list(range(2, self.n, 2))

    def step(self, psi):
        """One full time step of size dt. Mutates psi in place, returns it."""
        for b in self._odd:
            _apply_bond_gate(psi, b, self._gates_half[b], self.cutoff, self.maxdim)
        for b in self._even:
            _apply_bond_gate(psi, b, self._gates_full[b], self.cutoff, self.maxdim)
        for b in self._odd:
            _apply_bond_gate(psi, b, self._gates_half[b], self.cutoff, self.maxdim)
        return psi

File: src/dmrgpy/mpsjulialive/tebd.jl

# TEBD: 2nd-order Trotter (odd/even bond) real-time evolution for strictly
# nearest-neighbor Hamiltonians, run entirely in Julia. Julia counterpart of
# pyitensor/tebd.py (see that module's own docstring for the algorithm's
# full derivation and why it's correct for any parity-conserving fermionic
# Hamiltonian restricted to two adjacent sites) and mpscpp3/tebd.h (the C++
# port). One call per full nt-step trajectory, same design as tdvp.jl's
# quench_tdvp/evolve_and_measure_tdvp.
#
# Unlike tdvp_step (which hands a full Hamiltonian MPO to ITensorMPS's own
# tdvp()), TEBD needs each term's own *local* per-bond operator -- there is
# no way to slice that out of a compiled AutoMPO/MPO any more here than in
# C++ (automaton bookkeeping inflates even a single term's bond dimension;
# see tebd.h's own comment) -- so this file resolves each Hamiltonian term
# into its per-site matrices from scratch (read_terms/term_add!/
# resolve_term below), a direct port of pyitensor/autompo.py's
# HTerm.add()/HTerm.resolve().
#
# Deliberately fed *raw* term lists (mpo.py's text_mpo()/MO2list() shape,
# still using dmrgpy's own "C"/"Cdag" names) rather than
# MultiOperator.to_terms()'s Jordan-Wigner-*predressed* ("A"/"Adag"/"F")
# shape that mpscpp3/pyitensor's own TEBD consume: confirmed directly that
# real ITensors.jl's builtin "Fermion" site type
# (ITensors/src/lib/SiteTypes/src/sitetypes/fermion.jl) only defines
# op("C",..)/op("Cdag",..)/op("F",..) -- no "A"/"Adag" -- unlike dmrgpy's
# own hand-rolled pyitensor/C++ site tables, which do define "A"/"Adag" as
# dmrgpy's own name for the bare (string-free) annihilation/creation
# operator. ITensors.jl's op("C",s) *is* exactly that same bare, string-free
# operator (no JW string baked in -- the string only gets inserted by
# AutoMPO's automaton, or here, by term_add!/resolve_term below), so raw
# "C"/"Cdag" is both the form real ITensors.jl understands natively and the
# one dmrgpy already serializes for the existing MPO path (mpo.py's
# text_mpo/get_MPO) -- reused verbatim, via read_terms() below, rather than
# inventing a second serialization.
#
# is_fermionic_opname/term_add!/resolve_term operate on these raw names, so
# (unlike the pyitensor/C++ ports, where the equivalent carry-tracking logic
# is mostly an inert pass-through over already-JW-dressed input) the carry
# tracking here does the actual Jordan-Wigner threading.

using ITensors
using ITensorMPS
using LinearAlgebra


# Parses mpo.py's text_mpo()/read_operator.jl's read_operator() shared line
# format directly into a term list [(coef,[(name,site),...]),...] instead of
# compiling an AutoMPO -- same parsing loop as read_operator.jl's
# read_operator(), just building a plain list rather than an AutoMPO.
function read_terms(ls::Vector{String})
	nterms = parse(Int, ls[1])
	terms = Vector{Tuple{ComplexF64, Vector{Tuple{String, Int}}}}()
	for i in 1:nterms
		np = parse(Int, ls[2*i])
		l = split(ls[2*i+1], "  ")
		c = parse(Float64, l[1]) + im*parse(Float64, l[2])
		ops = Tuple{String, Int}[]
		for j in 1:np
			opname = String(l[2*j+1])
			ind = parse(Int, l[2*j+2])
			push!(ops, (opname, ind))
		end
		push!(terms, (c, ops))
	end
	return terms
end


# Name-based classification, mirrors ITensor's own isFermionic()
# (mpscpp3/ITensor/itensor/mps/autompo.cc) and pyitensor's
# sites/base.py::is_fermionic(): any operator name starting with 'C' is
# fermionic for Jordan-Wigner purposes.
is_fermionic_opname(name) = !isempty(name) && name[1] == 'C'


# Insertion-sort a term's raw (possibly site-unordered) factor list into
# site-ascending order, one factor at a time, picking up the sign from
# anticommuting a fermionic factor past every fermionic factor already fixed
# at a larger site -- direct port of pyitensor/autompo.py's HTerm.add().
# Mutates `ops` in place; `coefref` is a Ref so the running sign flip
# survives across repeated calls for the same term.
function term_add!(ops, coefref, opname, site)
	idx = 1
	while idx <= length(ops) && ops[idx][2] <= site
		idx += 1
	end
	if idx <= length(ops) && is_fermionic_opname(opname)
		nflip = count(o -> is_fermionic_opname(o[1]), ops[idx:end])
		if isodd(nflip)
			coefref[] *= -1
		end
	end
	insert!(ops, idx, (opname, site))
end


# The (dim,dim) matrix of single-site operator `opname` at site `s`, in the
# convention that plain matrix-vector multiplication reproduces op(opname,s)
# acting on a bare physical-index vector -- same convention/primitive as
# metts.jl's metts_site_operator_matrix.
function site_local_matrix(s, opname)
	O = op(opname, s)
	return Array(O, s', s)
end


# The full per-site list of (dim_i,dim_i) matrices (one per site, 1..n) that
# reconstructs this term's action once Kronecker-multiplied -- direct port
# of pyitensor/autompo.py's HTerm.resolve(): sites the term doesn't touch
# get plain Id, except while an odd number of fermionic operators have been
# "seen" scanning left to right, when they get F instead (threading the
# Jordan-Wigner string); a touched site composes its own factors (in their
# original relative order) and may pick up an extra trailing F of its own
# depending on the parity carried in from its left versus its own.
function resolve_term(sites, ops)
	n = length(sites)
	by_site = Dict{Int, Vector{String}}()
	for (name, site) in ops
		push!(get!(by_site, site, String[]), name)
	end
	out = Vector{Matrix{ComplexF64}}(undef, n)
	carry = false
	for site in 1:n
		s = sites[site]
		names = get(by_site, site, nothing)
		if names === nothing
			out[site] = site_local_matrix(s, carry ? "F" : "Id")
			continue
		end
		is_site_fermionic = isodd(count(is_fermionic_opname, names))
		need_F = (carry != is_site_fermionic)
		mats = [site_local_matrix(s, nm) for nm in names]
		if need_F
			push!(mats, site_local_matrix(s, "F"))
		end
		combined = mats[1]
		for m in mats[2:end]
			combined = combined * m
		end
		out[site] = combined
		carry = (carry != is_site_fermionic)
	end
	return out
end


# The 1-based (lo,hi) site range where `mats` differs from the local
# identity -- i.e. the sites this term actually, non-trivially acts on.
# `nothing` if the term is trivial everywhere (a bare scalar term, e.g. a
# ground-state-energy shift folded in as coef*Identity). Mirrors
# pyitensor/tebd.py's _true_span() -- see its own docstring for why this
# must be checked on the *resolved* matrices, not the raw op-name list.
function true_span(mats)
	touched = Int[]
	for (i, m) in enumerate(mats)
		d = size(m, 1)
		if !isapprox(m, Matrix{ComplexF64}(I, d, d); atol = 1e-8)
			push!(touched, i)
		end
	end
	isempty(touched) && return nothing
	return touched[1], touched[end]
end


# {b => ITensor} for b=1..n-1 (the bond between sites b and b+1), each an
# operator over (sites[b]',sites[b],sites[b+1]',sites[b+1]) -- built
# entirely from ITensor outer products (op(name,s), which already carries
# the correct (s',s) index orientation, and plain ITensor `*` for tensor-
# product placement onto a bond) rather than dense-matrix kron+reshape, so
# there is no manual index-ordering/array-layout bookkeeping to get wrong.
# One-site (onsite) terms are split half-and-half onto each of a site's
# neighboring bonds (full weight at a chain boundary) -- same convention as
# pyitensor/tebd.py's bond_hamiltonians(); a term spanning 3+ sites raises,
# since TEBD only supports strictly nearest-neighbor Hamiltonians.
function bond_hamiltonians(sites, terms)
	n = length(sites)
	if n < 2
		error("bond_hamiltonians: TEBD needs at least 2 sites")
	end
	H = Dict{Int, ITensor}()
	for b in 1:(n - 1)
		H[b] = ITensor(sites[b]', sites[b], sites[b+1]', sites[b+1])
	end
	for (coef, raw_ops) in terms
		ops = Tuple{String, Int}[]
		coefref = Ref(ComplexF64(coef))
		for (name, site) in raw_ops
			term_add!(ops, coefref, name, site)
		end
		c = coefref[]
		mats = resolve_term(sites, ops)
		span = true_span(mats)
		if span === nothing
			H[1] += c * (op("Id", sites[1]) * op("Id", sites[2]))
			continue
		end
		lo, hi = span
		if hi - lo > 1
			error("TEBD requires a nearest-neighbor Hamiltonian; found a " *
				  "term with non-trivial support spanning sites $lo..$hi")
		end
		if hi == lo
			has_left = lo > 1
			has_right = lo < n
			weight = (has_left && has_right) ? 0.5 : 1.0
			local_tensor = ITensor((weight * c) .* mats[lo], sites[lo]', sites[lo])
			if has_left
				b = lo - 1
				H[b] += op("Id", sites[b]) * local_tensor
			end
			if has_right
				b = lo
				H[b] += local_tensor * op("Id", sites[b+1])
			end
		else
			tA = ITensor(mats[lo], sites[lo]', sites[lo])
			tB = ITensor(mats[hi], sites[hi]', sites[hi])
			H[lo] += c * (tA * tB)
		end
	end
	return H
end


# The 2-site gate exp(-i*tau*Hbond) for one bond, via ITensors.jl's own
# tensor exponential (exp(A::ITensor), which auto-splits A's primed/
# unprimed indices into left/right) -- the Julia-native equivalent of
# mpscpp3/tebd.h's ITensor BondGate (also an ITensor-native exponential, not
# a manual expm) and pyitensor/tebd.py's scipy.linalg.expm (there, a manual
# dense exponential, needed because pyitensor has no ITensor-level exp()
# primitive of its own).
function bond_gate(Hbond, tau)
	return exp(-im * tau * Hbond)
end


# Contracts `gate` onto bond (i,i+1) and SVD-truncates, mutating psi in
# place. Canonicalizes psi's orthogonality center to i first (via
# orthogonalize!, the same primitive tdvp_step/entropy.jl's bond_entropy
# use), which is what makes the local SVD truncation-optimal regardless of
# which bond was last touched.
function apply_bond_gate!(psi, i, gate, cutoff, maxdim)
	orthogonalize!(psi, i)
	s_i = siteind(psi, i)
	theta = noprime(gate * psi[i] * psi[i+1], "Site")
	if i > 1
		left_link = commonind(psi[i], psi[i-1])
		U, S, V, _ = svd(theta, left_link, s_i; cutoff = cutoff, maxdim = maxdim)
	else
		U, S, V, _ = svd(theta, s_i; cutoff = cutoff, maxdim = maxdim)
	end
	psi[i] = U
	psi[i+1] = S * V
end


# Precomputes every bond's dt/2 (odd bonds) and dt (even bonds) evolution
# gate once from a static, nearest-neighbor Hamiltonian -- step() only ever
# needs those two, not both gates for every bond (2nd-order Trotter:
# exp(-i dt/2 H_odd) exp(-i dt H_even) exp(-i dt/2 H_odd)).
struct TEBDGates
	odd::Vector{Int}
	even::Vector{Int}
	gates_half::Dict{Int, ITensor}
	gates_full::Dict{Int, ITensor}
end

function build_tebd_gates(sites, terms, dt)
	n = length(sites)
	h_bonds = bond_hamiltonians(sites, terms)
	odd = collect(1:2:(n - 1))
	even = collect(2:2:(n - 1))
	gates_half = Dict(b => bond_gate(h_bonds[b], dt / 2.0) for b in odd)
	gates_full = Dict(b => bond_gate(h_bonds[b], dt) for b in even)
	return TEBDGates(odd, even, gates_half, gates_full)
end

function tebd_step!(psi, gates::TEBDGates, cutoff, maxdim)
	for b in gates.odd
		apply_bond_gate!(psi, b, gates.gates_half[b], cutoff, maxdim)
	end
	for b in gates.even
		apply_bond_gate!(psi, b, gates.gates_full[b], cutoff, maxdim)
	end
	for b in gates.odd
		apply_bond_gate!(psi, b, gates.gates_half[b], cutoff, maxdim)
	end
	return psi
end


# TEBD counterpart of tdvp.jl's quench_tdvp(): `ls` is a read_terms()-shaped
# text encoding of the (already energy-shifted) Hamiltonian's raw term list
# (mpo.py's text_mpo(), *not* an MPO); A1mpo/A2mpo/wf0 stay MPOs/MPS, same
# as quench_tdvp, since only the Hamiltonian needs per-term local gates --
# the observables are still applied via ordinary MPO application
# (apply_clean, reused unchanged from tdvp.jl).
function quench_tebd(ls, sites, A1mpo, A2mpo, wf0, nt, dt, cutoff, maxdim)
	terms = read_terms(ls)
	gates = build_tebd_gates(sites, terms, dt)
	psi1 = apply_clean(A1mpo, wf0, maxdim, cutoff)
	psi2 = apply_clean(A2mpo, wf0, maxdim, cutoff)
	norm0 = sqrt(abs(inner(psi1, psi1)))
	correlator = ComplexF64[]
	for it = 1:nt
		# Measure before evolving, so correlator[k] is C((k-1)*dt) and
		# lines up with timedependent.py's ts=[0,dt,...,(nt-1)*dt] grid --
		# see evolution_dmrg_DC()'s own comment there. Kept in lockstep with
		# the C++/pyitensor/ED copies of this loop.
		push!(correlator, inner(psi2, psi1))
		psi1 = tebd_step!(psi1, gates, cutoff, maxdim)
		psi1 = psi1 * (norm0 / norm(psi1))
	end
	return correlator, psi1
end


# TEBD counterpart of tdvp.jl's evolve_and_measure_tdvp() -- see that
# function's own comment for why renormalization here targets wf's own
# initial norm rather than 1. `wf` is the caller's own MPS object (unlike
# quench_tebd's psi1/psi2, which are always a fresh apply_clean() result),
# so it's copied before any mutation -- same reasoning as tdvp_step's own
# top-of-function copy().
function evolve_and_measure_tebd(ls, sites, Aop, wf, nt, dt, cutoff, maxdim)
	terms = read_terms(ls)
	gates = build_tebd_gates(sites, terms, dt)
	psi = copy(wf)
	norm0 = sqrt(abs(inner(psi, psi)))
	correlator = ComplexF64[]
	for it = 1:nt
		# Measure before evolving, so correlator[k] is C((k-1)*dt) and
		# lines up with timedependent.py's ts=[0,dt,...,(nt-1)*dt] grid --
		# see evolution_dmrg_DC()'s own comment there. Kept in lockstep with
		# the C++/pyitensor/ED copies of this loop.
		push!(correlator, inner(psi, Aop, psi))
		psi = tebd_step!(psi, gates, cutoff, maxdim)
		psi = psi * (norm0 / norm(psi))
	end
	return correlator, psi
end
